src/main/java/jenkins/plugins/logstash/persistence/BuildData.java

```diff
@@ -138,7 +139,7 @@ public TestData(Action action) {
   public BuildData(AbstractBuild<?, ?> build, Date currentTime, TaskListener listener) {
     initData(build, currentTime);
 
-    Node node = build.getBuiltOn();
+    Node node = build.getExecutor().getOwner().getNode();
```

Comment:
how about we try to use `getBuiltOn()` first and fall back to `getExecutor().getOwner().getNode()` if the first one is `null`? Just to be on the safe side

Reply:
It must be null if you follow the logic.
In the method _execute_ of class _Run_ you will see that the BuildWrappers are initialized from method _createBuildListener_. But the _builtOn_ is only set when later in the execute the _RunExecution.run_ is called (which actually is _AbstractBuild.AbstractBuildExecution.run_)

Reply:
I get that. However this is relying on implementation details. With the current low test coverage I'd prefer to be as conservative as possible.

Reply:
It is not pointless.
`getExecutor()` will only return something when a build is running. `getBuiltOn()` will also return the node when the build is finished when the node with this name still exists. Internally only the name of the node is stored and when you call `getBuiltOn()` it searches for the node with the name.
Maybe its time to write some tests with JenkinsRule where a real job is executed on a node.

Reply:
Thanks for the explanation.

Based on that I'd still prefer to keep the fallback logic in case somebody somewhere tries to pass a non-running build here.

> Maybe its time to write some tests with JenkinsRule where a real job is executed on a node.

Yeah this project definitely needs a heavy makeover (including this class).

I started to port the integ test we do on travis with docker in #36 but there is always something more urgent.
I also need to get more familiar with the jenkins test harness, hopefully #40 will be an opportunity to learn by example

Reply:
Another thing: how does this compare to the pipeline logic in lines 190-194? Seems to to the same thing but using `displayName`?

Reply:
This is wrong, it will return something like `Executor-#0`, but definitely not the host. 
For pipelines using `getExecutor().getOwner().getNode()` will just get the host where the logstashSend is called, but the build could have run on many different hosts before. 
Anyway I'm also looking into a more pipeline like logstash that accepts a block
```
logstash {
  node('label') {
    do something
    do anotherthing
  }
}
```
but I have to find a way to inject the host after the  logstashwriter is initialized.

Reply:
> will just get the host where the logstashSend is called, but the build could have run on many different hosts before.

I get that. 
One final question: for now, do you think we should change the pipeline handling to use the `getExecutor().getOwner().getNode()` instead of the current executor display name?

Reply:
I fixed also the pipeline handling to display the name of the node and not the executor. I also added some integration tests using JenkinsRule.